Discover JDKs during initialisation

Cy4Shot opened the pull request:
Added a step during intialisation to detect installed JDKs and their version

Cy4Shot committed:
Discover JDKs during initialisation (#87)

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+
+    public static void init() {
+        jdks = discoverJDKs();
+    }
```
I'm not a massive fan of this. I guess it works, it just doesn't feel like this is the most optimal way to achieve it. My personal recommendation, at least to start with, is:

```suggestion
    private static final List<JDK> jdks = new CopyOnWriteArrayList<>();

    public static void init() {
        jdks.addAll(discoverJDKs());
    }
```

Then afterwards, its worth considering a way for a user to define custom paths in which JDKs can live, or even directly add/remove a specific JDK. This is because otherwise this system is strictly limited to those locations.

It is also worth noting another common location, is `user.home/.jdks/`, as this is where IntelliJ often installs JDKs to.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+
+    private static String getJavaExecutable() {
+        return OperatingSystem.CURRENT == OperatingSystem.WINDOWS ? "java.exe" : "java";
+    }
```
This is a little bit deceiving to be a `get` method. Whilst technically yes it "gets" something, the common practise is to only name `get`/`set` when it actually "gets" or "sets" the data from/to somewhere. In this method thats not really the case, and therefore a better, more accurate method name might be something like `determineJavaExecutableName` or just `javaExecutableName`.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+    private static String findJavaOnPath() {
+        String pathEnv = System.getenv("PATH");
+        if (pathEnv == null) return null;
+        String[] paths = pathEnv.split(File.pathSeparator);
```
Just a minor formatting cleanup.
```suggestion
    private static String findJavaOnPath() {
        String pathEnv = System.getenv("PATH");
        if (pathEnv == null) 
            return null;

        String[] paths = pathEnv.split(File.pathSeparator);
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+        for (String p : paths) {
+            File javaFile = new File(p, JAVA_EXECUTABLE);
+            if (javaFile.exists() && javaFile.canExecute()) return javaFile.getAbsolutePath();
+        }
```
Consider using java.nio here rather than java.io.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+            File javaFile = new File(p, JAVA_EXECUTABLE);
+            if (javaFile.exists() && javaFile.canExecute()) return javaFile.getAbsolutePath();
+        }
+        return null;
```
```suggestion
        }

        return null;
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+            case LINUX -> LINUX_JDK_PATHS;
+            case UNKNOWN -> List.of();
+        };
+        return paths.stream().map(Paths::get).toList();
```
`Paths.get` is an old method that isn't needed anymore.
```suggestion
        return paths.stream().map(Path::of).toList();
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+            return new JDK(javaHome, name, version);
+        }
+        return null;
+    }
```
Should be using java.nio instead of java.io.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+        File f = new File(javaHomeOrPath, "bin/" + JAVA_EXECUTABLE);
+        if (f.exists() && f.canExecute()) {
+            javaExecutable = f.getAbsolutePath();
+        }
```
Should be using java.nio instead of java.io.
Also please name variables. `file` instead of `f` for example.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+        Map<String, JDK> uniqueJDKs = new LinkedHashMap<>();
+        for (JDK jdk : jdks) {
+            try {
+                String normalizedPath = Paths.get(jdk.path()).toRealPath().toString();
```
```suggestion
                String normalizedPath = Path.of(jdk.path()).toRealPath().toString();
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+            try {
+                String normalizedPath = Paths.get(jdk.path()).toRealPath().toString();
+                uniqueJDKs.putIfAbsent(normalizedPath, new JDK(normalizedPath, jdk.name(), jdk.version()));
+            } catch (IOException e) {
```
```suggestion
            } catch (IOException ignored) {
```

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,178 @@
+            }
+        } catch (IOException e) {
+            // Ignore
+        }
```
```suggestion
        } catch (IOException ignored) {}
```

DaRealTurtyWurty commented:
@Cy4Shot Another thing that needs to be pointed out is that this PR needs to target `dev`, not `main`.

DaRealTurtyWurty committed:
Add JDK management UI and brand detection

Introduces new settings panes for managing JDK scan paths and manual JDKs, adds detected JDK list UI, and enhances JDK brand detection with icons and images. Updates JDKManager to support exclusion and manual addition of JDKs, and adds new settings for JDK management. Integrates SVG support and updates build dependencies.

DaRealTurtyWurty committed:
Merge branch 'dev' into main

Signed-off-by: TurtyWurty <43997792+DaRealTurtyWurty@users.noreply.github.com>

YodaForce157 commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,439 @@
+        return filtered;
+    }
+
+    private static List<JDK> discoverJDKs() {
```
Javadocs on the more complex/logic intensive methods would be helpful, just something to explain what happens within the method when it's called would be good, as it can be useful for later debugging

YodaForce157 commented on src/main/java/dev/railroadide/railroad/java/JDKManager.java:
```diff
@@ -0,0 +1,439 @@
+                    }
+                }
+            }
+            case MAC -> MAC_JDK_PATHS.forEach(path -> addNormalized(candidates, Path.of(path)));
```
At the moment, mac support seems to be broken, as I have jdk 21 installed and it just says there is a distribution at path /usr, with a distribution type of usr...
I will look into this and figure out why this is, however, the JDKManager should always attempt to avoid providing non-existent JDKS

DaRealTurtyWurty replied:
I've done a bit of a rework of the Mac stuff (and discovery in general), and with a bit of research plus some guidance from gemini, I believe it should work now.

YodaForce157 replied:
Ok so, it can now detect the jdk correctly, however it puts the Distribution as Home, due to the path being something like /Library/Java/JavaVirtualMachines/ms-21.0.7/Contents/Home

DaRealTurtyWurty replied:
Yeah, I don't know how that could be easily fixed tbh. Best way is probably to make this a specific edge case condition and it will just have to go up 2 directories to determine the distribution if the path ends with `/Contents/Home`.

YodaForce157 commented on src/main/java/dev/railroadide/railroad/settings/SettingCodecs.java:
```diff
@@ -65,4 +69,22 @@ public class SettingCodecs {
             .jsonDecoder(KeybindsList::fromJson)
             .build();
+
+    public static final SettingCodec<List<Path>, DirectoryListPane> DIRECTORY_PATH_LIST =
```
I don't really like this at all, I think it's more of a fault of the settings manager, but having 3 large sections for something this simple is too much. Ideally, I personally would prefer a section that had a dropdown (each section in settings should have a dropdown, but again, this is a settings manager issue) and then a table that holds a path and an enum variable that holds either EXCLUDE or INCLUDE, the path should be able to be differentiated between a file and dir by just checking for a / at the end, this will be checked when it's inputted into the table

YodaForce157 replied:
Same thing for detected JDKS really, and it would be nice to allow the user to add an alias or name for each detected jdk just for some QoL

DaRealTurtyWurty replied:
Yeah I agree, its not ideal, but really the settings system needs rewriting first.

DaRealTurtyWurty replied:
> Same thing for detected JDKS really, and it would be nice to allow the user to add an alias or name for each detected jdk just for some QoL

The only issue with this is that the detected JDKs aren't ever stored. And it would kinda defeat the point if they were to be, so I don't necessarily know how that would work. Because for them to be stored, currently it'd have to live in the settings, as we don't have any sort of persistent storage API.

YodaForce157 reviewed (commented):
Would be great if you could add some javadocs to most if not all the methods in the JDKManager class.
Will need to fix macos support (I'll look into it) and eventually fix the settings but that needs the settings system fixed soo

YodaForce157 commented:
Ok, so looking into the issue with mac, it appears to call createJDKFromAnyPath("/usr/bin/java"), which then called resolveJavaHome("/usr/bin/java") which returns just /usr. This then returns a new JDK with just /usr. Funnily enough, I also do not have a /usr/bin/java directory.
Seems to be resolveJavaHome causing the issue, specifically the if statement at line 363

DaRealTurtyWurty committed:
Refactor JDK discovery logic and add JDKUtils helper

Moved JDK discovery and parsing utilities from JDKManager to a new JDKUtils class for better separation of concerns. Updated JDKManager to use JDKUtils and FileUtils for path normalization and JDK metadata extraction. Improved platform-specific JDK detection, added more robust version parsing, and enhanced comments for clarity. Minor UI logic fix in DetectedJdkListPane for icon handling and version formatting.

DaRealTurtyWurty commented on src/main/java/dev/railroadide/railroad/settings/ui/SettingsPane.java:
```diff
@@ -125,11 +125,20 @@ public SettingsPane() {
-            ));
+            );
+
+            // TODO: Temporary until we add a decorations system.
+            if ("ide".equals(parts[parts.length - 1])) {
+                var detectedPane = new DetectedJdkListPane();
+                VBox.setMargin(detectedPane, new Insets(10, 10, 0, 10));
+                vbox.getChildren().add(detectedPane);
+            }
```
Can't be changed until the settings system has been modified, which is out of scope of this PR.

DaRealTurtyWurty committed:
Implement copilot suggestions

- Check for vendor and implementor as first port of call
- Scan only C-Z drives as A and B are floppy only.
- Add setting for java version checking timeout
- Minor formatting cleanups
